src/Google.Api.Gax.Grpc/BufferedClientWriter.cs

```diff
@@ -0,0 +1,285 @@
+        public Task TryWriteAsync(T message) => WriteAsyncImpl(message, null, false, false);
+
+        /// <summary>
+        /// Writes a message to the stream, if there is enough space in the buffer and <see cref="CompleteAsync"/>
+        /// hasn't already been called.
+        /// </summary>
+        /// <param name="message">The message to write.</param>
+        /// <param name="options">The write options to use for this message.</param>
+        /// <returns><c>null</c> if the message queue is full or the stream has already been completed.</returns>
+        public Task TryWriteAsync(T message, WriteOptions options) => WriteAsyncImpl(message, options, true, false);
+
+        /// <summary>
+        /// Writes a message to the stream, if there is enough space in the buffer and <see cref="CompleteAsync"/>
+        /// hasn't already been called. The same write options will be used as for the previous message.
+        /// </summary>
+        /// <param name="message">The message to write.</param>
+        /// <exception cref="InvalidOperationException">There isn't enough space left in the buffer,
+        /// or the stream has been completed.</exception>
+        /// <returns>A <see cref="Task"/> which will complete when the message has been written to the stream.</returns>
+        public Task WriteAsync(T message) => WriteAsyncImpl(message, null, false, true);
+
+        /// <summary>
+        /// Writes a message to the stream, if there is enough space in the buffer and <see cref="CompleteAsync"/>
+        /// hasn't already been called.
+        /// </summary>
+        /// <param name="message">The message to write.</param>
+        /// <param name="options">The write options to use for this message.</param>
+        /// <exception cref="InvalidOperationException">There isn't enough space left in the buffer,
+        /// or <see cref="CompleteAsync"/> has already been called.</exception>
+        /// <returns>A <see cref="Task"/> which will complete when the message has been written to the stream.</returns>
+        public Task WriteAsync(T message, WriteOptions options) => WriteAsyncImpl(message, options, true, true);
+
+        private Task WriteAsyncImpl(T message, WriteOptions options, bool modifyOptions, bool throwOnError)
+        {
+            GaxPreconditions.CheckNotNull(message, nameof(message));
+            lock (_lock)
+            {
+                if (!ValidateStateForWrite(throwOnError))
+                {
+                    return null;
+                }
+                // Not part of state validation, as it's slightly different.
+                if (_failedTask != null)
+                {
+                    return _failedTask;
+                }
+                var lastRequest = _queue.Last?.Value;
+                var effectiveOptions = modifyOptions ? options
+                    : lastRequest != null ? lastRequest.WriteOptions
```

Comment:
Nope - because if `lastRequest?.WriteOptions` is null, that's what we want to propagate. null is a valid value.